ibc/store/provable_store.go

```diff
@@ -0,0 +1,244 @@
+	var proof *ics23.CommitmentProof
+	if membership {
+		proof, err = p.CreateMembershipProof(key, value)
+	} else {
+		proof, err = p.CreateNonMembershipProof(key)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	return value, proof, nil
+}
+
+// CreateMembershipProof creates a membership proof for the key-value pair with the key
+// prefixed with the CommitmentPrefix, by importing the state tree from the TreeStore
+func (p *provableStore) CreateMembershipProof(key, value []byte) (*ics23.CommitmentProof, error) {
+	// import IBC state tree
+	// TODO(#854): Implement tree retrieval
+	/**
+	prefixed := applyPrefix(p.prefix, key)
+	root, nodeStore := p.bus.GetTreeStore().GetTree(trees.IBCStateTree)
+	lazy := smt.ImportSparseMerkleTree(nodeStore, root, sha256.New())
+	return createMembershipProof(lazy, prefixed, value)
+	**/
+	return nil, nil
+}
+
+// CreateNonMembershipProof creates a non-membership proof for the key prefixed with the
+// CommitmentPrefix, by importing the state tree from the TreeStore
+func (p *provableStore) CreateNonMembershipProof(key []byte) (*ics23.CommitmentProof, error) {
+	// import IBC state tree
+	// TODO(#854): Implement tree retrieval
+	/**
+	prefixed := applyPrefix(p.prefix, key)
+	root, nodeStore := p.bus.GetTreeStore().GetTree(trees.IBCStateTree)
+	lazy := smt.ImportSparseMerkleTree(nodeStore, root, sha256.New())
+	return createNonMembershipProof(lazy, prefixed)
+	**/
+	return nil, nil
+}
+
+// Set updates the persistence layer with the new key-value pair at the latest height and
+// emits an UpdateIBCStore event to the bus for propagation throughout the network, to be
+// included in each node's mempool and thus the next block
+func (p *provableStore) Set(key, value []byte) error {
+	prefixed := applyPrefix(p.prefix, key)
+	currHeight := int64(p.bus.GetConsensusModule().CurrentHeight())
+	rwCtx, err := p.bus.GetPersistenceModule().NewRWContext(currHeight)
+	if err != nil {
+		return err
+	}
+	defer rwCtx.Release()
+	if err := rwCtx.SetIBCStoreEntry(prefixed, value); err != nil {
+		return err
+	}
+	p.cache = append(p.cache, &cachedEntry{
+		storeName:   p.name,
+		height:      uint64(currHeight),
+		prefixedKey: prefixed,
+		value:       value,
+	})
+	// TODO(#854): Implement emit functions
+	// return emitUpdateStoreEvent(p.prefix, key, value)
+	return nil
+}
+
+// Delete updates the persistence layer with the key and nil value pair at the latest height
+// and emits an PruneIBCStore event to the bus for propagation throughout the network, to be
+// included in each node's mempool and thus the next block
+func (p *provableStore) Delete(key []byte) error {
+	prefixed := applyPrefix(p.prefix, key)
+	currHeight := int64(p.bus.GetConsensusModule().CurrentHeight())
+	rwCtx, err := p.bus.GetPersistenceModule().NewRWContext(currHeight)
+	if err != nil {
+		return err
+	}
+	defer rwCtx.Release()
+	if err := rwCtx.SetIBCStoreEntry(prefixed, nil); err != nil {
+		return err
+	}
+	p.cache = append(p.cache, &cachedEntry{
```

Comment:
IMPORTANT COMMENT.

What if someone sends multiple `Store` and `Prune` operations for the same key throughout the duration of a single block?

I believe the behaviour (and expected outcome) is arguably unexpected. My suggestion is to update the `cache` to be a map. Consider even looking at `GenericFIFOSet` and `txFIFOMempool` of how its used.

The outcome is still non-deterministic (it is subdued to race conditions) but I think its a better approach